### PR TITLE
Firefox reorg

### PR DESCRIPTION
**Pull request by jdungan**
Get out your GIFs, this is one mother of a PR  ... but you get the new firefox addon.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -1,39 +1,40 @@
...
+1. Run 'gulp dev-chrome'   
+2. Follow these instructions (http://developer.chrome.com/extensions/getstarted.html#unpacked) 
+3. Point to the unpacked codesy chrome extension in ./chrome.
+4. Reload the extension page (chrome://extensions/)
+
+
+To use the Firefox Addon
+------------------------
+1. Run 'gulp dev-firefox'   
+2. Run 'jpm run --profile default'  (this runs a new instance of firefox with your default profile so your bookmarks, etc. are available)
```
Comment (groovecoder) [inline README.rst]: You have to run `jpm` from the `firefox` directory. So this should be:

1. `gulp dev-firefox`
2. `cd firefox`
3. `jpm run --profile default`

File: README.rst [hunk 2]
```diff
@@ -1,39 +1,40 @@
...
+--------------------------------------
+1. Run 'gulp dev-chrome'   
+2. Follow these instructions (http://developer.chrome.com/extensions/getstarted.html#unpacked) 
+3. Point to the unpacked codesy chrome extension in ./chrome.
+4. Reload the extension page (chrome://extensions/)
+
+
+To use the Firefox Addon
+------------------------
+1. Run 'gulp dev-firefox'
```
Comment (groovecoder) [inline README.rst]: Hmm ... when I ran this it seemed to hang?
```
^C(codesy)lcrouch-14912:chrome-extension lcrouch$ gulp dev-firefox
[20:42:17] Using gulpfile ~/code/codesy/chrome-extension/gulpfile.js
[20:42:17] Starting 'load-static'...
[20:42:17] Finished 'load-static' after 6.63 ms
[20:42:17] Starting 'firefox-package'...
[20:42:17] Finished 'firefox-package' after 2.19 ms
[20:42:17] Starting 'firefox-coffee'...
[20:42:17] Finished 'firefox-coffee' after 107 ms
[20:42:17] Starting 'dev-firefox'...
[20:42:17] Finished 'dev-firefox' after 7.56 ms
```

File: README.rst [hunk 3]
```diff
@@ -1,39 +1,40 @@
...
 2. Set up Gulp
-
    * node install gulp
    * 'npm install' to install dependencies
+3. For firefox run 'npm install jpm'
+
+
+To use the 'unpacked' Chrome Extension
+--------------------------------------
+1. Run 'gulp dev-chrome'
```
Comment (groovecoder) [inline README.rst]: This also hung and never returned ...
```
^C(codesy)lcrouch-14912:chrome-extension lcrouch$ gulp dev-chrome
[20:43:25] Using gulpfile ~/code/codesy/chrome-extension/gulpfile.js
[20:43:25] Starting 'load-static'...
[20:43:25] Finished 'load-static' after 6.18 ms
[20:43:25] Starting 'chrome-manifest'...
[20:43:25] Finished 'chrome-manifest' after 2.61 ms
[20:43:25] Starting 'chrome-coffee'...
[20:43:25] Finished 'chrome-coffee' after 84 ms
[20:43:25] Starting 'dev-chrome'...
[20:43:25] Finished 'dev-chrome' after 7.19 ms
```

File: README.rst [hunk 4]
```diff
@@ -1,39 +1,40 @@
...
-
-
 Development
 -----------
 1. Clone this repo
 2. Set up Gulp
-
    * node install gulp
    * 'npm install' to install dependencies
+3. For firefox run 'npm install jpm'
```
Comment (groovecoder) [inline README.rst]: We may as well add `jpm` to the `package.json` so a single `npm install` will install everything. Installing this way is only for devs anyway.

File: README.rst [hunk 5]
```diff
@@ -1,39 +1,40 @@
...
-   (You can also run gulp dev-start)   
-3. Follow these instructions (http://developer.chrome.com/extensions/getstarted.html#unpacked)
-
-
 Development
 -----------
 1. Clone this repo
 2. Set up Gulp
-
    * node install gulp
```
Comment (groovecoder) [inline README.rst]: We can remove this line too. `gulp` is already listed in `package.json`

File: README.rst (hunk 3 above)
Comment (groovecoder) [inline README.rst]: Duh, I'm an idiot. It started watching.

File: README.rst (hunk 2 above)
Comment (groovecoder) [inline README.rst]: This also started watching.

File: README.rst [hunk 6]
```diff
@@ -1,39 +1,40 @@
...
-4. Load the unpacked extension (see above)
-5. Make changes the coffee script files in ./src and the ./prod/manifest.json files
-6. Reload the extension page (chrome://extensions/)
 
 
-When you are ready to publish
------------------------------
+When you are ready to publish (Chrome only -- FIX THIS)
+-------------------------------------------------------
 1. Run 'gulp publish'
```
Comment (groovecoder) [inline README.rst]: This is now `gulp publish-chrome`

File: package.json [hunk 7]
```diff
@@ -1,38 +1,39 @@
...
+    "name": "codesy-chrome-extension", 
+    "repository": {
+        "url": "https://github.com/jdungan/chrome-extension.git", 
+        "type": "git"
+    }, 
+    "author": "John Dungan", 
+    "bugs": {
+        "url": "https://github.com/jdungan/chrome-extension/issues"
+    }, 
+    "version": "0.0.0",
```
Comment (groovecoder) [inline package.json]: Is it worth keeping the Chrome and Firefox versions in sync? I supposed in theory one could need a bug-fix that the other doesn't?

Comment (groovecoder) [inline package.json]: Bah, this is the npm version.

File: src/firefox/package.json [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+  "id": "codesy@codesy.io",
+  "description": "codesy is a pay-what-you-want market for the open source community to encourage coders to fix important bugs.",
+  "author": "Luke Crouch",
+  "license": "MPL 2.0",
+  "version": "0.0.9",
+  "main" : "./data/js/index.js",
+  "permissions": {
+    "cross-domain-content": [
+      "https://api.codesy.io/",
+      "https://localhost:8443/"
```
Comment (groovecoder) [inline src/firefox/package.json]: Take the `localhost` permission out of the `package.json` so the Firefox add-on only asks for production permission? (Devs will have to add it back to communication with localhost, but we can document that.)

File: src/issue.coffee [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+          codesy.append err.responseText
+        else
+          console.log err
+
+codesy.urlChange = () ->
+  if codesy.href isnt window.location.href
+    codesy.href = window.location.href
+    codesy.getAuth()
+  
+  window.setTimeout codesy.urlChange, 600
```
Comment (groovecoder) [inline src/issue.coffee]: cheating

**Comment by groovecoder** [comment]
The code is looking good. Why are we back to including pure CSS in the static assets of the widgets?

And I couldn't get `jpm run --profile default` to run with the Firefox add-on. :cry:

File: src/firefox/package.json (hunk 8 above)
Comment (jdungan) [inline src/firefox/package.json]: We can add it to the gulp-firefox task.  I skipped this just so there would be some technical debt to work on.

**Comment by groovecoder** [comment]
So, this works when I use `jpm run` without a profile.

**Comment by jdungan** [comment]
The CSS had to be added back because Firefox enforces github's CSP:

Content Security Policy: The page's settings blocked the loading of a resource at https://127.0.0.1:8443/static/img/icon48.png ("img-src https://github.com data: https://assets-cdn.github.com https://identicons.github.com https://www.google-analytics.com https://collector.githubapp.com https://*.githubusercontent.com https://*.gravatar.com https://*.wp.com").
Content Security Policy: The page's settings blocked the loading of a resource at https://127.0.0.1:8443/static/css/styles.css ("style-src https://github.com 'unsafe-inline' 'unsafe-eval' https://assets-cdn.github.com").
Content Security Policy: The page's settings blocked the loading of a resource at https://cdn.jsdelivr.net/pure/0.5.0/pure-min.css ("style-src https://github.com 'unsafe-inline' 'unsafe-eval' https://assets-cdn.github.com").

**Comment by groovecoder** [comment]
Not sure why the profile wasn't running with the extension before. It seems to be running now with a new profile ...

**Comment by groovecoder** [comment]
Seems the codesy add-on is incompatible with Firefox 36+ ? Does this still work for you on Firefox 36?

**Comment by jdungan** [comment]
Fixed issue where signing in or out with FF was not recognized by active widgets.  Now all widgets register a callback for receiving changes.

Tested fine on Firefox 36

**Comment by groovecoder** [comment]
I got `jpm` to work. I had to downgrade to version `0.0.23` ([apparently there's a bug in latest versions](https://github.com/mozilla/mdn-devtool/issues/11#issuecomment-76745007))

I sent https://github.com/jdungan/chrome-extension/pull/1.

I still get CSP errors though. :frowning: 

firefox-csp-widget

**Pull request merged**